Objective-C/CBLDictionary.m

```diff
@@ -0,0 +1,444 @@
+        return $castIf(CBLArray, value);
+}
+
+
+- (BOOL) containsObjectForKey: (NSString*)key {
+    id value = _dict[key];
+    if (!value)
+        return [super containsObjectForKey: key];
+    else
+        return value != kRemovedValue ? YES : NO;
+}
+
+
+- (NSArray*) allKeys {
+    NSMutableSet* result = [NSMutableSet setWithArray: [_dict allKeys]];
+    for (NSString* key in [super allKeys]) {
+        if (![result containsObject: key])
+            [result addObject: key];
+    }
+    
+    [_dict enumerateKeysAndObjectsUsingBlock:^(NSString *key, id value, BOOL *stop) {
+        if (value == kRemovedValue)
+            [result removeObject: key];
+    }];
+    
+    return [result allObjects];
+}
+
+
+- (NSDictionary<NSString*,id>*) toDictionary {
+    NSMutableDictionary* result = [_dict mutableCopy];
+    
+    // Backing data:
+    NSDictionary* backingData = [super toDictionary];
+    [backingData enumerateKeysAndObjectsUsingBlock: ^(id key, id value, BOOL *stop) {
+        if (!result[key])
+            result[key] = value;
+    }];
+    
+    for (NSString* key in [result allKeys]) {
+        id value = result[key];
+        if (value == kRemovedValue)
+            result[key] = nil; // Remove key
+        else if ([value conformsToProtocol: @protocol(CBLReadOnlyDictionary)])
+            result[key] = [value toDictionary];
+        else if ([value conformsToProtocol: @protocol(CBLReadOnlyArray)])
+            result[key] = [value toArray];
+    }
+    
+    return result;
+}
+
+
+#pragma mark - SETTER
+
+
+- (void) setObject: (nullable id)value forKey: (NSString*)key {
+    id oldValue = [self objectForKey: key];
+    if (!$equal(value, oldValue)) {
+        value = [self prepareValue: value];
+        [self detachChangeListenerForObject: oldValue];
+        [self setValue: value forKey: key isChange: YES];
+    }
+}
+
+
+- (void) setDictionary: (NSDictionary<NSString*,id>*)dictionary {
+    // Detach all objects that we are listening to for changes:
+    [self detachChildChangeListeners];
+    
+    NSMutableDictionary* result = [NSMutableDictionary dictionary];
+    [dictionary enumerateKeysAndObjectsUsingBlock: ^(id key, id value, BOOL *stop) {
+        result[key] = [self prepareValue: value];
+    }];
+    
+    // Marked the key as removed by setting the value to kRemovedValue:
+    NSDictionary* backingData = [super toDictionary];
+    [backingData enumerateKeysAndObjectsUsingBlock: ^(id key, id value, BOOL *stop) {
+        if (!result[key])
+            result[key] = kRemovedValue;
+    }];
+    
+    _dict = result;
+    
+    [self setChanged];
+}
+
+
+- (void) removeObjectForKey: (NSString*)key {
+    [self setObject: nil forKey: key];
+}
+
+
+#pragma mark - SUBSCRIPTION
+
+
+- (CBLFragment*) objectForKeyedSubscript: (NSString*)key {
+    id value = [self objectForKey: key];
+    return [[CBLFragment alloc] initWithValue: value parent: self parentKey: key];
+}
+
+
+#pragma mark - INTERNAL
+
+
+- (BOOL) isEmpty {
```

Comment:
The logic here is wrong. Consider where the superclass has an FLDict containing only the key 'x', and this object's `_dict` maps 'x' to 'kRemovedValue'. This method will incorrectly return `NO`.